### ISSUE TITLE
Source loss 관련하여...

### ISSUE THREAD

**Issue by cjs6211** [postBG/DTA.pytorch#8]
#Train 시 

            # Source CE Loss
            source_features1, source_features2 = self.feature_extractor(source_inputs)
            source_logits1, source_logits2 = self.classifier(source_features1), self.classifier(source_features2)

            # Source pi model
            ce_loss = self.class_criterion(source_logits1, source_labels)
            source_consistency_loss = 2 * self.source_cnn_consistency_weight * self.source_consistency_criterion(
                source_logits2, source_logits1)
            source_loss = ce_loss + source_consistency_loss
            source_loss.backward()

에서 source_logit1, source_logits2는 완벽하게 같지 않나요?
그럼 self.source_consistency_criterion(source_logits2, source_logits1)은 의미가 없을 것 같은데요...

**Comment by postBG** [comment]
안녕하세요. @cjs6211님.
self.feature_extractor는 내부에 dropout layer를 포함하고 있습니다.
source_feature1과 source_feature2는 같은 입력 이미지에 대해 서로 다른 dropout을 적용하여 만든 feature입니다. 따라서 source_logits1과 source_logits2도 서로 다른 값이 됩니다.

**Comment by cjs6211** [comment]
self.feature_extractor 는 ResNetLower이고 이 클래스의 forward 출력을 보면 

h1 = self.layer4(x)
h2 = self.layer4(x)
        return h1, h2

이렇게 나와있습니다. tariner의 train 함수(train_one_epoch)에서는 target image에 대해 dropout mask를 적용하고 있는 것이 아래의 코드를 통해 확인되고 있습니다. 

target_logits_cnn_drop = self.classifier(target_cnn_dropout_mask * target_features2)
target_logits_fc_drop = self.classifier(target_features2, target_fc_dropout_mask)
target_consistency_loss = self.target_consistency_criterion(target_logits_cnn_drop, target_logits1)
target_consistency_loss +=self.target_consistency_criterion(target_logits_fc_drop, target_logits1)

하지만 source image에 대해서는 적용되고 있지 않은 것 같습니다. 사실 논문에서는 source image에 대해 ce loss만 취급하셔서 source에 대해 굳이 consistency loss를 개입시킬 필요는 없습니다만...

**Comment by postBG** [comment]
안녕하세요. self.layer4에 [드랍아웃](https://github.com/postBG/DTA.pytorch/blob/ca33b3f118c3fcb1a2b9f6bfaf57681180e206b2/models/resnet.py#L87)이 포함되어 있습니다.
저희가 [Adversarial Dropout](https://arxiv.org/abs/1707.03631)의 구현 트릭을 사용하기 때문입니다.

실제로
```python
import torch
from models.visda_architectures import create_resnet_lower

feature_extractor = create_resnet_lower()
f1, f2 = feature_extractor(torch.rand(1, 3, 224, 224))
print(f1.equal(f2))
```
를 실행해보면 False가 나옵니다.

말씀하신대로 source consistency loss이 성능에 큰 영향을 미치지 않는 것이 사실이기는 합니다.

**Comment by cjs6211** [comment]
감사합니다. 많은 도움이 되었습니다. 

> 안녕하세요. self.layer4에 [드랍아웃](https://github.com/postBG/DTA.pytorch/blob/ca33b3f118c3fcb1a2b9f6bfaf57681180e206b2/models/resnet.py#L87)이 포함되어 있습니다.
> 저희가 [Adversarial Dropout](https://arxiv.org/abs/1707.03631)의 구현 트릭을 사용하기 때문입니다.
> 
> 실제로
> 
> ```python
> import torch
> from models.visda_architectures import create_resnet_lower
> 
> feature_extractor = create_resnet_lower()
> f1, f2 = feature_extractor(torch.rand(1, 3, 224, 224))
> print(f1.equal(f2))
> ```
> 
> 를 실행해보면 False가 나옵니다.
> 
> 말씀하신대로 source consistency loss이 성능에 큰 영향을 미치지 않는 것이 사실이기는 합니다.
